api/handlers.go

```diff
@@ -117,3 +143,96 @@ func (api *API) parseAndResolve(skylink string) (skymodules.Skylink, error) {
 	}
 	return sl, nil
 }
+
+// threadedPerformSweep performs the actual sweep operation.
+// TODO This can be moved to a separate `sweeper` type which would also hold the
+//  status of the latest sweep as well as the relevant mutex. This can
+//  streamline the structure of API. I recommend this as a F/U.
+func (api *API) threadedPerformSweep() {
+	api.latestSweepStatusMu.Lock()
+	// Double-check for parallel sweeps.
+	if api.latestSweepStatus.InProgress {
+		api.latestSweepStatusMu.Unlock()
+		return
+	}
+	// Initialise the status to "a sweep is running".
+	api.latestSweepStatus = SweepStatus{
+		InProgress: true,
+		Error:      nil,
+		StartTime:  time.Now().UTC(),
+		EndTime:    time.Time{},
+	}
+	api.latestSweepStatusMu.Unlock()
+	// Define an error variable which will represent the success of the scan.
+	var err error
+	// Ensure that we'll finalize the sweep on returning from this method.
+	defer func() {
+		api.latestSweepStatusMu.Lock()
+		api.latestSweepStatus.InProgress = false
+		api.latestSweepStatus.EndTime = time.Now().UTC()
+		api.latestSweepStatus.Error = err
+		api.latestSweepStatusMu.Unlock()
+	}()
+
+	// Perform the actual sweep.
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	var cacheErr error
+	go func() {
+		defer wg.Done()
+		res := api.staticSkydClient.RebuildCache()
+		<-res.Ch
+		cacheErr = res.ExternErr
+	}()
+
+	// We use an independent context because we are not strictly bound to a
+	// specific API call. Also, this operation can take significant amount of
+	// time and we don't want it to fail because of a timeout.
+	ctx := context.Background()
+	dbCtx, cancel := context.WithDeadline(ctx, time.Now().UTC().Add(database.MongoDefaultTimeout))
+	defer cancel()
+
+	// Get pinned skylinks from the DB
+	dbSkylinks, err := api.staticDB.SkylinksForServer(dbCtx, api.staticServerName)
+	if err != nil {
+		err = errors.AddContext(err, "failed to fetch skylinks for server")
+		return
+	}
+	wg.Wait()
+	if cacheErr != nil {
+		err = errors.AddContext(cacheErr, "failed to revuild skyd cache")
+		return
+	}
+
+	unknown, missing := api.staticSkydClient.DiffPinnedSkylinks(dbSkylinks)
+
+	// Remove all unknown skylink from the database.
+	var s skymodules.Skylink
+	for _, sl := range unknown {
+		s, err = database.SkylinkFromString(sl)
+		if err != nil {
+			err = errors.AddContext(err, "invalid skylink found in DB")
+			build.Critical(err)
+			return
```

Comment:
I agree. I was thinking "critical == stop what you are doing" but there isn't a need to do that.
I'll change those to `continue`s. I'll keep the other errors as `return`s because they will likely be caused by something more general like a DB failure.